### ISSUE TITLE
Invalid evhtp_connection_free results segfault with SSL

### ISSUE THREAD

**Issue by t-shinde** [criticalstack/libevhtp#94]
Version: libevhtp-1.2.10 I am using old version but i think same problem is there with new version also.
Details:
I am not sure if following is exact reason, but this is what I found after debugging
1. evhtp accept connection and wait for enough data and feed it to htparser_run
2. htparser_run may call default callback which this request will get served from this function. 
I mean we are in _evhtp_connection_readcb --> htparser_run --> _evhtp_request_parser_fini
3. Note the first line of _evhtp_connection_readcb is  evhtp_connection_t * c = arg;
4. In parser_fini, if user calls evhtp_send_reply<start|body|end> it will do bufferevent_write_buffer(c->bev, buf); on very same connection from which we came
5. bufferevent_write_buffer ->evbuffer_add_buffer->evbuffer_invoke_callbacks(outbuf); This invoke callbacks hits error, and calls _evhtp_connection_eventcb - which does evhtp_connection_free((evhtp_connection_t *)arg);
This mean if we return to (2) via stack unwinding connection memory is stale. this results in segfault

==22965== Invalid read of size 8
==22965==    at 0x70606A1: _evhtp_connection_readcb (evhtp.c:1833)
==22965==    by 0x70A3900: _bufferevent_run_readcb (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70B2E35: consider_reading (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70B31E2: be_openssl_readeventcb (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70963B0: event_persist_closure (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70964BF: event_process_active_single_queue (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x709678C: event_process_active (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x7096DF4: event_base_loop (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x7068F18: _evthr_loop (evthr.c:126)
==22965==    by 0x6E39E24: start_thread (pthread_create.c:308)
==22965==    by 0x6B6734C: clone (clone.S:113)
==22965==  Address 0x1fed9570 is 112 bytes inside a block of size 160 free'd
==22965==    at 0x4C2ACBD: free (vg_replace_malloc.c:530)
==22965==    by 0x70643EE: evhtp_connection_free (evhtp.c:3966)
==22965==    by 0x7060AB5: _evhtp_connection_eventcb (evhtp.c:1971)
==22965==    by 0x70A3A9A: _bufferevent_run_eventcb (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70B24DB: conn_closed (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70B2B3C: do_write (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70B2FC4: consider_writing (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70B3911: be_openssl_outbuf_cb (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x709CD93: evbuffer_run_callbacks (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x709CE8A: evbuffer_invoke_callbacks (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x709E383: evbuffer_add_buffer (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70A3E6A: bufferevent_write_buffer (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==  Block was alloc'd at
==22965==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==22965==    by 0x7060DE6: _evhtp_connection_new (evhtp.c:2080)
==22965==    by 0x706100A: _evhtp_accept_cb (evhtp.c:2166)
==22965==    by 0x70A6F6B: listener_read_cb (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70963B0: event_persist_closure (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x70964BF: event_process_active_single_queue (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x709678C: event_process_active (in /home/tushar/inst/opt/extern/lib/libevhtp.so)
==22965==    by 0x7096DF4: event_base_loop (in /home/tushar/inst/opt/extern/lib/libevhtp.so)

**Comment by NathanFrench** [comment]
Greetings!!

What version of libevent and ssl are you using?

Is there some code i can use to reproduce?

**Comment by t-shinde** [comment]
libevent-devel-2.0.21-4.el7.x86_64
openssl-libs-1.0.2k-8.el7.x86_64
openssl-1.0.2k-8.el7.x86_64
openssl-devel-1.0.2k-8.el7.x86_64

I had possible fix 
`diff --git a/org/libevhtp-1.2.10/evhtp.h b/libevhtp-1.2.10/evhtp.h`
`index 4e3d3ea..9d0d348 100644`
`--- a/org/libevhtp-1.2.10/evhtp.h`
`+++ b/libevhtp-1.2.10/evhtp.h`
`@@ -153,7 +153,7 @@ typedef void (* evhtp_hook_err_cb)(evhtp_request_t * req, evhtp_error_flags errt`
` typedef void (* evhtp_hook_event_cb)(evhtp_connection_t * conn, short events, void * arg);`
` `
` /* Generic hook for passing ISO tests */`
`-typedef evhtp_res (* evhtp_hook)();`
**`+typedef evhtp_res (* evhtp_hook)(void);`**
` `
` typedef evhtp_res (* evhtp_hook_conn_err_cb)(evhtp_connection_t * connection, evhtp_error_flags errtype, void * arg);`
` typedef evhtp_res (* evhtp_pre_accept_cb)(evhtp_connection_t * conn, void * arg);`
`@@ -434,8 +434,12 @@ struct evhtp_request_s {`
` };`
` `
` #define evhtp_request_content_len(r) htparser_get_content_length(r->conn->parser)`
**`+#define evhtp_connection_ref(c) (__sync_add_and_fetch(&((c)->ref_cnt), 1))`
`+#define evhtp_connection_cnt(c) (__sync_add_and_fetch(&((c)->ref_cnt), 0))`
`+#define evhtp_connection_unref(c) (__sync_add_and_fetch(&((c)->ref_cnt), -1))`**
` `
` struct evhtp_connection_s {`
**`+    uint64_t   ref_cnt;`**
`     evhtp_t  * htp;`
`     evbase_t * evbase;`
`     evbev_t  * bev;`
`diff --git a/org/libevhtp-1.2.10/evhtp.c b/libevhtp-1.2.10/evhtp.c`
`index 65e0485..9ba073b 100644`
`--- a/org/libevhtp-1.2.10/evhtp.c`
`+++ b/libevhtp-1.2.10/evhtp.c`
`@@ -1800,11 +1800,15 @@ _evhtp_connection_readcb(evbev_t * bev, void * arg) {`
`         return;`
`     }`
` `
**`+    evhtp_connection_ref(c);`
`+`**
`     buf   = evbuffer_pullup(bufferevent_get_input(bev), avail);`
` `
`     nread = htparser_run(c->parser, &request_psets, (const char *)buf, avail,`
`                       c->htp->vrts_mx_ct_len);`
` `
**`+    evhtp_connection_unref(c);`
`+`**
`     if (c->owner != 1) {`
`         /*`
`          * someone has taken the ownership of this connection, we still need to`
`@@ -1835,6 +1839,8 @@ _evhtp_connection_readcb(evbev_t * bev, void * arg) {`
`     } else if (htparser_get_error(c->parser) != htparse_error_none) {`
`         evhtp_connection_free(c);`
`     }`
`+`
`+    return;`
` } /* _evhtp_connection_readcb */`
` static void`
`@@ -3934,6 +3940,10 @@ evhtp_connection_free(evhtp_connection_t * connection) {`
`         return;`
`     }`
` `
**`+    if (evhtp_connection_cnt(connection)) {`
`+          return;`
`+    }`
`+`**
`     _evhtp_connection_fini_hook(connection);`
`     evhtp_safe_free(connection->request, _evhtp_request_free);`
` `

I am not sure about this fix. But idea is if we are using connection_t in htparser_run we should not allow it to be freed. and hence do ref counting on it. I am sure it this not "Complete" fix and there will be other places where we should inc/dec refcount. Please share your thoughts on this.

**Comment by t-shinde** [comment]
test code to explain above fix.
[test.c.gz](https://github.com/criticalstack/libevhtp/files/1939041/test.c.gz)
wget --no-check-certificate https://s3.earth:8143/hello

**Comment by NathanFrench** [comment]
TL;DR: please test the latest releases before you attempt to fix a problem that may have already been solved.

I was unfortunately unable to reproduce this with the current release of libevent and libevhtp.

I have one erk with any attempts at pull requests. And that erk is atomic operations.

Commonly thought of as silver-bullets alternative to, and more performant than mutexes, etc. There are a few reasons for my overall distaste for atomics, obviously the first is that this is a gcc'ism. Not portable, and is inlined for specific architectures (i.e., `-march=i386` will produce very different results from `-march=[pentium|native]`).

Personally I'm not opposed to such things, but I've had flack in the past for using atomic intrinsics in thread.c.

Things have to be aligned correctly - you did one half right here, you made it uint64, one of 2, 4, 8; a new gap introduced, and an 8 bytes reference counter? *no thanks*.

Not to mention in order to work optimally, we would need to introduce even more gcc'isms like `__align__(())` on structures - and as of right now, they are aligned perfectly for both 32b and 64b (and 32K/64K cache).

One of the little snippets of goodness that came along between 10 and 16 was a header only API for [transparent refcounts](https://github.com/criticalstack/libevhtp/blob/develop/refcount.h). Something that you may want to look into for this. You can also setup your own custom allocators and all the "normal" things you can do with other api's. They also trickle down to the libevent level.

There have been so many changes between the current version and the version you are using, I would rather point you to the "best-of" changelogs in the releases dir: https://github.com/criticalstack/libevhtp/releases this even includes openssl-1.x fixes which you will need as the abi has changed drastically.

-- 

But I will continue looking into this, but confined in the current release. In the mean time - if you do decide to try out a newer release, and it doesn't work either : I'll send you along an amazon gift card after it's fixed :)

**Comment by t-shinde** [comment]
I tried to update to latest build on version 1.2.16 branch master.
First problem is if we use ipv6: as bind address, server only listen on IPv6 sockets and not on IPv4 - unlike previous version. it took me a while to understand following lines changed the behavior and not my network :)

[root@sun_01 git-libevhtp]# git diff
diff --git a/evhtp.c b/evhtp.c
index 6eaf319..f506c98 100644
--- a/evhtp.c
+++ b/evhtp.c
@@ -4078,6 +4078,7 @@ evhtp_bind_sockaddr(evhtp_t * htp, struct sockaddr * sa, size_t sin_len, int bac
             break;
         }
 
+#if 0
         if (sa->sa_family == AF_INET6)
         {
             if (setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, &on, sizeof(on)) == -1)
@@ -4085,6 +4086,7 @@ evhtp_bind_sockaddr(evhtp_t * htp, struct sockaddr * sa, size_t sin_len, int bac
                 break;
             }
         }
+#endif
 
         if (bind(fd, sa, sin_len) == -1)
         {
@@ -4094,7 +4096,6 @@ evhtp_bind_sockaddr(evhtp_t * htp, struct sockaddr * sa, size_t sin_len, int bac
         error = 0;
     } while (0);
 
-
     if (error == 1)
     {
         if (fd != -1)

[root@sun_01 git-libevhtp]# git branch
  develop
* master
[root@sun_01 git-libevhtp]# 


1. Is there reason why we want to force IPv6 only in this case? If yes how can I enable server to bind on any ipv4 or ipv6 connection?
2. Is master stable branch ? i.e. what nomenclature you follow for stable branches?

I will test this new lib and share results.

**Comment by NathanFrench** [comment]
> Is there reason why we want to force IPv6 only in this case? If yes how can I enable server to bind on any ipv4 or ipv6 connection?

Dual-stacked binds can be problematic unless it's specifically designated as either v4 or v6.

To dual stack an `ANY` in the examples section you can find the following:

```C
    struct event_base * evbase;
    evhtp_t           * htp_v6;
    evhtp_t           * htp_v4;
    int                 rc;

    evbase = event_base_new();

    htp_v6 = evhtp_new(evbase, NULL);
    htp_v4 = evhtp_new(evbase, NULL);

    evhtp_set_gencb(htp_v6, _req_cb, (void *)"ipv6");
    evhtp_set_gencb(htp_v4, _req_cb, (void *)"ipv4");

    evhtp_bind_socket(htp_v6, "ipv6:::/128", 9090, 1024);
    evhtp_bind_socket(htp_v4, "ipv4:0.0.0.0", 9090, 1024);

    event_base_loop(evbase, 0);
```


> Is master stable branch ? i.e. what nomenclature you follow for stable branches?

We follow the `git-flow` model.

`develop` is always deemed semi-stable, but not production ready. 
`master` is pre-release, and the step right before becoming a release. Nothing new is ever pushed here until directly. That is until a pre-release is marked. 

In the releases section, there are two types: pre-release, and release, which are both marked as such. As of the time of writing https://github.com/criticalstack/libevhtp/releases/tag/1.2.16 is the current release.

Thank you very much!
